### PR TITLE
web: support SAML resource deletion in unified resources view

### PR DESCRIPTION
**Pull request by flyinghermit**
Changes summary:
- Delete menu button is added.
- `ResourceAction` type is added that specifies `edit` or `delete` action.
- Checks user permission to SAML IdP service provider to disable `edit` or `delete` menu. 
- Removes SAML app from the unified resource view once the resource item is deleted from the backend. 
- Refactor's edit flow to use context.
  - `SamlAppActionContext` has been added to provide edit and delete actions for Menu buttons. The `SamlAppActionProvider` added in this PR does nothing and only satisfies the `useSamlAppAction` hook. Working provider has been added to the `e` PR.

Part of https://github.com/gravitational/teleport.e/issues/4458

### CODE REVIEW

**Comment by ravicious** [comment]
I left a review under the teleport.e PR, I'd like to understand why OSS needs to know about `setResourceAction`. https://github.com/gravitational/teleport.e/pull/4673#pullrequestreview-2182988956

File: web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx [hunk 1]
```diff
@@ -186,6 +192,20 @@ export function ClusterResources({
...
+    const index = resources.findIndex(e => {
+      switch (e.kind) {
+        case 'app':
+          return e.name === samlAppToDelete.name && e.samlApp;
+        default:
+          return false;
+      }
+    });
+    if (index >= 0) {
+      resources.splice(index, 1);
```
Comment (ravicious) [inline web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx:206]: `Array.prototype.splice` mutates the array. In React, you almost never want to mutate state. https://react.dev/learn/updating-objects-in-state#local-mutation-is-fine

Instead, you can filter it instead:

<details>
<summary>Patch</summary>

```diff
diff --git a/web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx b/web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx
index 472bbca42c..cc49985411 100644
--- a/web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx
+++ b/web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx
@@ -16,7 +16,7 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-import React, { useCallback, useState } from 'react';
+import React, { useCallback, useMemo, useState } from 'react';
 
 import { Flex } from 'design';
 import { Danger } from 'design/Alert';
@@ -155,7 +155,12 @@ export function ClusterResources({
     getClusterPinnedResources: getCurrentClusterPinnedResources,
   };
 
-  const { fetch, resources, attempt, clear } = useUnifiedResourcesFetch({
+  const {
+    fetch,
+    resources: unfilteredResources,
+    attempt,
+    clear,
+  } = useUnifiedResourcesFetch({
     fetchFunc: useCallback(
       async (paginationParams, signal) => {
         const response = await teleCtx.resourceService.fetchUnifiedResources(
@@ -192,20 +197,22 @@ export function ClusterResources({
       ]
     ),
   });
+
   const { samlAppToDelete } = useSamlAppAction();
-  if (samlAppToDelete?.backendDeleted) {
-    const index = resources.findIndex(e => {
-      switch (e.kind) {
-        case 'app':
-          return e.name === samlAppToDelete.name && e.samlApp;
-        default:
-          return false;
-      }
-    });
-    if (index >= 0) {
-      resources.splice(index, 1);
-    }
-  }
+  const resources = useMemo(
+    () =>
+      samlAppToDelete?.backendDeleted
+        ? unfilteredResources.filter(
+            res =>
+              !(
+                res.kind === 'app' &&
+                res.samlApp &&
+                res.name === samlAppToDelete.name
+              )
+          )
+        : unfilteredResources,
+    [samlAppToDelete, unfilteredResources]
+  );
 
   // This state is used to recognize when the `params` value has changed,
   // and reset the overall state of `useUnifiedResourcesFetch` hook. It's tempting to use a

```
</details>

Comment (flyinghermit) [reply web/packages/teleport/src/UnifiedResources/UnifiedResources.tsx]: True, I definitely overlooked this one from poc to PR updated with your patch in https://github.com/gravitational/teleport/pull/44311/commits/774df9aa5f3c55150628661d0d5d25b7d9a48dec

File: web/packages/shared/components/UnifiedResources/CardsView/ResourceCard.story.tsx [hunk 2]
```diff
@@ -136,20 +130,22 @@ export const Cards: Story = {
...
-            selected={false}
-            pinningSupport={PinningSupport.Supported}
-            name={res.name}
-            primaryIconName={res.primaryIconName}
-            SecondaryIcon={res.SecondaryIcon}
-            cardViewProps={res.cardViewProps}
-            labels={res.labels}
-            ActionButton={res.ActionButton}
-          />
+          <SamlAppActionProvider>
```
[the remark this answers is not in the document]
Comment (flyinghermit) [reply web/packages/shared/components/UnifiedResources/CardsView/ResourceCard.story.tsx]: good catch - https://github.com/gravitational/teleport/pull/44311/commits/774df9aa5f3c55150628661d0d5d25b7d9a48dec

**Comment by flyinghermit** [comment]
@gzdunek @avatus could I also get your review on this one please. The PR has gone through some architectural changes but is ready for review again.

**Review by avatus — APPROVED** [review]

**Review by ravicious — APPROVED** [review]

**Pull request merged**